### ISSUE TITLE
Question: Running a command in a pod?

### ISSUE THREAD

**Issue by bradleyd** [coryodaniel/k8s#29]
Hi, I want to be able to run a command in a pod and this requires a websocket upgrade when using `Core.V1.connect_get_namespaced_pod_exec`. Is there any examples you can point me at ?

Thanks

**Comment by coryodaniel** [comment]
`connect` isn't current supported :(

**Comment by coryodaniel** [comment]
Definitely open to a PR, but this has been blocked behind. #12

**Comment by coryodaniel** [comment]
Hacking something together may be possible with [this](https://github.com/coryodaniel/k8s/blob/master/docs/usage.md#performing-unsupported-sub-resource-operations) and [this library](https://github.com/Azolo/websockex)

If you come up with anything that could be used for `/attach` or `/exec` send me a PR <3

**Comment by bradleyd** [comment]
Ok thanks.

**Comment by Rodeoclash** [comment]
I have used https://github.com/Azolo/websockex to consume logs from a container in Kubernetes, had some issues with the use of CA certs which I never got to the bottom of, but aside from that it works without issue.

**Comment by bradleyd** [comment]
Yeah, I am running into that same issue right now =]

**Comment by Rodeoclash** [comment]
You can construct an implementation of the WebSockex server where the `conn` struct is customised, here's a very stripped back version of what I have in my app:
```elixir
defmodule MyApp.K8sLogging do
  use WebSockex

  def conn() do
    WebSockex.Conn.new(
      url,
      cacerts: [your_cert]
    )
  end

  def start_link([preview]) do
    WebSockex.start_link(
      conn(),
      __MODULE__
    )
  end

  def handle_frame({_type, message}, %{preview: preview} = state) do
    IO.inspect(message)
  end
end
```
Might be a useful starting point for you

**Comment by bradleyd** [comment]
Thanks! I got pretty far but I think the certs are the culprit. I will keep tinkering and share what I have got. Maybe even get a PR =]

`{:error, %WebSockex.RequestError{code: 406, message: "Not Acceptable"}}`

**Comment by coryodaniel** [comment]
Nice. Heads up #12 is WIP for adding sub resource support so this might time nicely for 0.3

**Comment by bradleyd** [comment]
ok. Looking at why I am getting unauthorized using my certs. Something to work on over the holiday. 
Thanks!

**Comment by bradleyd** [comment]
Here is a contrived example (WIP) of Pod exec working. I tried this against a local multipass microk8s cluster and our dev cluster in AWS. 
I will put something more formal together after I look at #12 . I wanted to get it here in case you are further along and can use this for an example.

This is using the default config for `~/.kube/config`

``` elixir
defmodule PodExec do
  use WebSockex

  def start_link(%{name: name, ns: ns, command: command, container: container} = state) do
    operation = K8s.Operation.build(:get, "v1", :pod, namespace: ns, name: name)

    with {:ok, base_url} <- K8s.Cluster.url_for(operation, :default),
         {:ok, cluster_connection_config} <- K8s.Cluster.conf(:default),
         {:ok, request_options} <- K8s.Conf.RequestOptions.generate(cluster_connection_config) do
      exec_url =
        "#{base_url}/exec?#{command_builder(command, [])}container=#{container}&stdin=true&stderr=true&stdout=true&tty=true"

      headers =
        request_options.headers ++ [{"Accept", "*/*"}, {"Content-Type", "application/json"}]

      cacerts = Keyword.get(request_options.ssl_options, :cacerts)

      conn =
        WebSockex.Conn.new(exec_url,
          insecure: false,
          ssl_options: request_options.ssl_options,
          cacerts: cacerts,
          extra_headers: headers
        )

      WebSockex.start_link(conn, __MODULE__, state)
    end
  end

  def start_link(%{name: name, ns: ns, command: command} = state) do
    operation = K8s.Operation.build(:get, "v1", :pod, namespace: ns, name: name)

    with {:ok, base_url} <- K8s.Cluster.url_for(operation, :default),
         {:ok, cluster_connection_config} <- K8s.Cluster.conf(:default),
         {:ok, request_options} <- K8s.Conf.RequestOptions.generate(cluster_connection_config) do
      exec_url =
        "#{base_url}/exec?#{command_builder(command, [])}stdin=true&stderr=true&stdout=true&tty=true"

      headers =
        request_options.headers ++ [{"Accept", "*/*"}, {"Content-Type", "application/json"}]

      cacerts = Keyword.get(request_options.ssl_options, :cacerts)

      conn =
        WebSockex.Conn.new(exec_url,
          insecure: false,
          ssl_options: request_options.ssl_options,
          cacerts: cacerts,
          extra_headers: headers
        )
      WebSockex.start_link(conn, __MODULE__, state)
    end
  end

  # stdout
  def handle_frame({type, <<1, msg::binary>>}, state) do
    IO.puts("Received STDOUT Message - Type: #{inspect(type)} -- Message: #{inspect(msg)}")
    {:ok, state}
  end

  # stderr
  def handle_frame({type, <<2, msg::binary>>}, state) do
    IO.puts("Received STDERR Message - Type: #{inspect(type)} -- Message: #{inspect(msg)}")
    {:ok, state}
  end

  def handle_frame({type, msg}, state) do
    IO.puts("Received Unknown Message - Type: #{inspect(type)} -- Message: #{inspect(msg)}")
    {:ok, state}
  end

  def handle_cast({:send, {type, msg} = frame}, state) do
    IO.puts("Sending #{type} frame with payload: #{msg}")
    {:reply, frame, state}
  end

  defp command_builder([], acc), do: Enum.reverse(acc)
  defp command_builder([h | t], acc) do
    command_builder(t, ["command=#{h}&" | acc])
  end
end
```

```elixir
iex(1)> PodExec.start_link(%{name: "nginx-8586bbc6b5-tlnxm", ns: "default", command: ["date"]})
{:ok, #PID<0.614.0>}
Received STDOUT Message - Type: :binary -- Message: ""
Received STDOUT Message - Type: :binary -- Message: ""
Received STDOUT Message - Type: :binary -- Message: "Wed Jul 10 21:33:16 UTC 2019\r\n"
```

**Comment by coryodaniel** [comment]
That's rad, my friend.

I'm a few test cases away from finishing up the WIP referenced above. Was hoping to work on it today, but got caught up in some stuff.

This is a great start!

**Comment by coryodaniel** [comment]
Shipped v0.3 w/ support for subresources, so this should be unblocked. Sorry for the delay, life and all that jazz.

**Comment by Rodeoclash** [comment]
Awesome stuff!

I've been keeping a list of things I found complicated or challenging with the implementation of your library, my intention is to (eventually!) submit a PR to your usage guide with some additions around how to do certain things. A good example is passing query params to select resources based on say a `labelSelector`. It took a bit of digging through the source to figure out that it's an argument given to the `run` method under `Client`.

**Comment by bradleyd** [comment]
No worries.  I have been behind as well.
Hopefully put something together soon.

Brad

**Comment by Rodeoclash** [comment]
Getting a bit of strange behaviour around consuming the sub resources, in particular getting the logs for a container in a pod. Here's a dump of the variables being passed to HTTPoison:

```elixir
iex(2)> App.Previews.KubernetesV2.Templates.Pod.logs({p}, "idfb142617986e4-6c8ccc4798-gx2cs") |> Application.get_env(:app, :kubernetes_v2).run([params: %{"container" => "preview"}])
%K8s.Operation{
  api_version: "v1",
  data: nil,
  method: :get,
  name: "pods/log",
  path_params: [
    namespace: "idfb142617986e4",
    name: "idfb142617986e4-6c8ccc4798-gx2cs"
  ],
  verb: :get
}
"https://192.168.99.108:8443/api/v1/namespaces/idfb142617986e4/pods/idfb142617986e4-6c8ccc4798-gx2cs/log"

> {:ok, nil}
```

Picking apart the URL that's actually being passed to HTTPoison, it looks correct `https://192.168.99.108:8443/api/v1/namespaces/idfb142617986e4/pods/idfb142617986e4-6c8ccc4798-gx2cs/log?container=preview` (and matches the same URL I get if I ask for the logs via `kubectl`)

Digging to the response, the logs request will return a string body not in JSON format, e.g.

```elixir
{:ok,
  %HTTPoison.Response{
    body: "--- Pulling\nUsing default tag: latest\n"
    ...
```

Interestingly, this doesn't raise when parsing to JSON but instead returns `nil`. What are your thoughts about only parsing the JSON depending on response type and passing the string responses back directly? Kubernetes helpfully tells us the payload is `{"Content-Type", "application/json"}` for a JSON response or `{"Content-Type", "text/plain"}` for the logs response so we could conditionally parse in https://github.com/coryodaniel/k8s/blob/52863a8438ef9f612dfae790d33736f18da55999/lib/k8s/client/http_provider.ex#L59

**Comment by coryodaniel** [comment]
@Rodeoclash can you open an new issue. I think that sounds like a good solution. Do you think it shoudl return a list of lines or a blob of the whole response?

**Comment by bradleyd** [comment]
FWIW here is a first pass at adding this. [PR](https://github.com/coryodaniel/k8s/pull/68)

**Comment by aleDsz** [comment]
Hey, any news about this feature? Would be awesome to use this feature to execute some code inside a pod using Elixir.
